### ISSUE TITLE
Acceptance-corrected Yield in the plotter

### ISSUE THREAD

**Issue by keigo-miz** [mashephe/AmpTools#111]
To evaluate the fit quality (for example, for SDME analyses), measured distributions and MC distributions weighted with fit results are compared as seen in FIG. 6 in 
https://halldweb.jlab.org/DocDB/0055/005576/015/Paper_Draft__rho_SDMEs_2233003.pdf

But at least for this rho-meson analysis, the distributions are determined mainly by acceptance effects and it's very hard to tell whether the fit is really good or not by just seeing this comparison.
I think the comparison between the acceptance-corrected yield and the model is somehow better when it comes to the evaluation of the fit.
Here,
* acceptance = (#accepted MC events weighted with fit results) / ($generated MC events weighted with fit results)
* the model = generated MC event distribution weighted with fit results

For example, in the plot c) of FIG. 6 in the above rho-meson paper draft, this comparison gives an impression that the fit works well because the discrepancy looks small. But in fact, when you check the acceptance-corrected yield vs the model, clear systematic discrepancies are observed.

In short, I think it's better to check this acceptance-corrected yield ( = ( data - background ) / acceptance ) in the plotter, and what is the easiest way to achieve this?

I'd like to hear your opinions and raise this feature request if this is not implemented.
Thanks!

Keigo Mizutani

**Comment by mashephe** [comment]
Hi Keigo,

In general it is a difficult/impossible problem to correct data for acceptance.  The problem is that the acceptance is a non-trivial function of the multidimensional phase space that the amplitudes are written in.  The only way to determine this function would be to bin phase space and determine the acceptance in each bin -- this is usually not feasible.  In addition if there are zeros or small numbers in the acceptance then the acceptance-corrected data is not well defined.

The fit formalism is really built on the idea of generating p.d.f. for the observed events from a model.  So it makes sense to compare this p.d.f. with the observed events, and that is what is being done in the plots in Fig. 6.  The plots in Fig. 6 do have a limitation:  they integrate over other variables and project onto one variable.  There can be disagreements in the multi-dimensional space that get washed out in this integration.  This is a general problem that has no easy solution.  At best you can explore two variables at a time and compare contours of the model with a scatter plot of data.  This capability is supported in the plotting framework, but it is difficult to do visually.  The plots in Fig. 6 are useful:  agreement of model with data is a necessary condition for a quality fit.  It is not a sufficient condition because the general problem is in a high dimensional space.

Matt

**Comment by keigo-miz** [comment]
Hi Matt and all,

Thank you for the reply.
I agree that the acceptance is a non-trivial function of the multidimensional space, but still it can be calculated using the weights (or the fit results) as follows:
* acceptance = (#accepted MC events weighted with fit results) / (#generated MC events weighted with fit results)

For example, when you calculate the acceptance distribution as a function of phi_hel [c) in Fig. 6], what you need is to take the ratio of the following distributions:
* accepted MC phi_hel distribution (weighted)
* generated MC phi_hel distribution (weighted)
And both distributions are actually what we already have in the plotter.
(Of course this distribution is an integrated one-dimensional acceptance, but correlation effects are taken into account by using the fit results (=weights).)

And after obtaining the acceptance, the acceptance-corrected yield can be calculated by
* acceptance-corrected yield = ( data - background ) / acceptance
and this distribution is what should be compared to "generated MC distribution weighted with fit results" (= the model).

Mathematically, this comparison (acceptance-corrected yield vs the model) is the same as that is already done by "data vs weighted MC" comparison. But my point is that the impressions are different, at least, for c) in Fig. 6.
In c) of Fig. 6, the distribution shapes are mainly determined by the acceptance effects. Even if you change the model parameters (SDMEs, in this case), let's say by 0.1, it's almost impossible to detect how the distribution shapes has been changed by eye.
On the other hand, by checking the "acc-corrected yield vs the model", you can easily see how the model (or the model parameters) works, and you can even imagine what a certain parameter value should be just by looking at the distributions.
(In case of c) in Fig. 6, this distribution is sensitive to rho^0_{1-1}.)

The distribution that I'd like to check is actually straightforwardly obtained from what we already have in the plotter. I just think it's worth to implement the feature that users can easily check this "acceptance-corrected yield" distribution in the plotter.

Any comments and opinions are welcome. Thanks!
Keigo

**Comment by mashephe** [comment]
> On Apr 7, 2023, at 11:05 AM, keigo-miz ***@***.***> wrote:
> 
> And after obtaining the acceptance, the acceptance-corrected yield can be calculated by
> 
> 	• acceptance-corrected yield = ( data - background ) / acceptance
> and this distribution is what should be compared to "generated MC distribution weighted with fit results" (= the model).
> Mathematically, this comparison (acceptance-corrected yield vs the model) is the same as that is already done by "data vs weighted MC" comparison. B

I think this is not correct.  In your equation above you are writing the numerator and denominator as integrals over the other dimension of phase space that you are not plotting.  It is not sufficient to divide the integral of the data by the integral of the acceptance.  One really needs to know how the data populate the other dimensions ... and the acceptance in those specific regions, not just the average acceptance.

Matt

**Comment by keigo-miz** [comment]
> In your equation above you are writing the numerator and denominator as integrals over the other dimension of phase space that you are not plotting. 

Yes, this is true. But still, given the data is well modeled in the whole multi-dimensional space, I think this "integrated" or "averaged" acceptance is correct (because it's 'correctly' integrated/averaged) and it's worth to compare "acc-corrected yield" and "the model" above.
The plots in Fig. 6 are okay and are worth checking. But checking the acc-corrected yield, we can more easily identify in which region of a certain variable (like phi_hel), data is deviated from the model.
I'm not saying this is sufficient, I just mean this can be done in a straightforward manner from what we already have in the plotter, and it's worth checking (maybe this part you won't agree..).

Keigo

**Comment by mashephe** [comment]
> On Apr 7, 2023, at 11:52 AM, keigo-miz ***@***.***> wrote:
> 
> Yes, this is true. But still, given the data is well modeled in the whole multi-dimensional space, I think this "integrated" or "averaged" acceptance is correct (because it's 'correctly' integrated/averaged) and it's worth to compare "acc-corrected yield" and "the model" above.
> The plots in Fig. 6 are okay and are worth checking. But checking the acc-corrected yield, we can more easily identify in which region of a certain variable (like phi_hel), data is deviated from the model.

If you do this, you don't know if data is deviating from the model or the acceptance that you used to do the correction was incorrect or the background is not a faithful representation of the true background.  Fig. 6(c) says there is a mismatch between the fit and the data, but you don't know what is causing it.  

I strongly prefer to make the minimum number of mathematical operations.  I think the plots like Fig. 6 are most close to what the fitting algorithm is doing internally.  There is no notion of an acceptance corrected yield anywhere in the context of the fit.

All that said, there is nothing to keep you from generating the plots that you describe.  The PlotGenerator framework that the GUI uses can deliver all of the components you need and you can manipulate them as you wish using standard TH1F::Divide, etc..  I encouraged most analysts to write their own short piece of code for plotting all the diagnostic histograms relevant for an analysis.  (This is much more efficient if you don't need the interactive features of the GUI.)  In such a piece of code one can do the manipulations.  If you make some of these plots and they appear to be useful or reveal new information, then we can consider a standard option in the GUI.

Matt

**Comment by keigo-miz** [comment]
Thank you. I understand and agree with your preference that the number of mathematical operations should be minimum.
And also, thank you for explaining how I can make/manipulate histograms.

Keigo

**Comment by mashephe** [comment]
I meant to put a link to the example code -- if you want to work with the histograms, you may look here:

https://github.com/mashephe/AmpTools/blob/master/Tutorials/Dalitz/DalitzExe/plotResults.cc

Around line 130 is where this code starts working with a TH1F pointer that you could use for other manipulations.  If you need any help deciphering the intermediate lines, let me know and I'm happy to help.

Matt

**Comment by mashephe** [comment]
With this discussion I'm not aware of any specific change needed so I'm closing this issue.  It can be reopened if there is new information.
